File: 113-path-sum-ii/113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> res;
    void dfs(TreeNode* root,int k,vector<int>ans){
        if(root==NULL ) return ;
        ans.push_back(root->val);
        k-=root->val;
        if(root->left==NULL and root->right==NULL){
            if(k==0){
                res.push_back(ans);
            }
            return;
        }
        dfs(root->left,k,ans);
        dfs(root->right,k,ans);
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        res.clear();
        vector<int>ans;
        dfs(root,targetSum,ans);
        return res;
    }
};

File: 113-path-sum-ii/113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        ans=[]
        self.dfs(root,targetSum,ans,[]);
        return ans;
    
    
    def dfs(self,root,k,ans,res):
        if root==None:
            return
        k-=root.val
        if root.left ==None and root.right==None and k==0:
            ans.append(res+[root.val])
            return
        self.dfs(root.left,k,ans,res+[root.val])
        self.dfs(root.right,k,ans,res+[root.val])